temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowInternal.java

```diff
@@ -539,7 +541,41 @@ public static <R> R getLastCompletionResult(Class<R> resultClass, Type resultTyp
     return getRootWorkflowContext().getLastCompletionResult(resultClass, resultType);
   }
 
-  public static void upsertSearchAttributes(Map<String, Object> searchAttributes) {
+  @Nullable
+  public static <T> T getSearchAttribute(String name) {
+    List<T> list = getSearchAttributeValues(name);
+    if (list == null) {
+      return null;
+    }
+    Preconditions.checkState(list.size() > 0);
+    Preconditions.checkState(
+        list.size() == 1,
+        "search attribute with name '%s' contains a list '%s' of values instead of a single value",
+        name,
+        list);
+    return list.get(0);
+  }
+
+  @Nullable
+  public static <T> List<T> getSearchAttributeValues(String name) {
+    SearchAttributes searchAttributes = getRootWorkflowContext().getContext().getSearchAttributes();
+    if (searchAttributes == null) {
+      return null;
```

Comment:
I wonder if this should return an empty list for no value. Won't `getSearchAttributes` map values be empty lists instead of null?

It seems we return null in all cases of no attributes, no attribute, and no attribute value. So they can't use nullability to differentiate set or not anyways. If not, I haven't checked, but we probably need to state in the javadoc that this can never return an empty list then.

Reply:
Well, we use an empty list as a "delete" token, so I would day it's safer to deserialize it always to null to users.

> If not, I haven't checked, but we probably need to state in the javadoc that this can never return an empty list then.
👍